### PR TITLE
perf: Improve Multiple Chunk Upload Performance

### PR DESCRIPTION
**Pull request by danielbankhead**
`chunkSize` uploads should be much faster.  🦕

### CODE REVIEW

**Review by danielduhh — COMMENTED** [review]
This looks like a significant change to the resumable write path. I suggest a design with some background information & comprehensive performance tests before merging this

File: src/resumable-upload.ts
Comment (ddelgrosso1) [inline src/resumable-upload.ts:267]: I think you mentioned removing this in our discussion but does encoding actually ever get used for anything?

Comment (danielbankhead) [reply src/resumable-upload.ts:267]: Not really; only for incoming buffers that are encoded to strings - which we decode here:

https://github.com/googleapis/nodejs-storage/blob/efb99b8aad6dbb8cf698cfcabc6a4d1fc3bbefd8/src/resumable-upload.ts#L398

**Comment by ddelgrosso1** [comment]
Overall I think this looks good and since we walked through the code I understand the why/how. I think once the supporting document is a bit more fleshed out with some metrics, I am good with this getting merged / released.

**Review by ddelgrosso1 — APPROVED** [review]
Going to approve this pending the results of perf tests.

**Pull request merged**
